DATAES-672 - Introduce SearchHit and SearchHits types to enrich search results

sothawo opened the pull request:
This Draft-PR  introduces `SearchHit<T>` which will be returned from the `SearchOperations` methods instead of a plain `T`. It currently has as additional information the score of the hit, more stuff like highlights, sort etc. will be added later with new issues for that.

Repositories don't return these as it would be impossible to distinguish at runtime of a repository method should return `List<T>` or `List<SearchHit<T>>`. This will be adresses in a separate issue to enable this for repositories by introducing `SearchHits<T>`.

This is a draft PR as these changes are in the non-reactive part currently. I will add the `SearchHit` to the reactive part in the next days probably.

mp911de commented:
Spring Data MongoDB has a bit of code to distinguish between `List<T>`, `List<GeoResult<T>>` and `GeoResults<T>`, see [`AbstractMongoQuery`](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/AbstractMongoQuery.java#L133-L142) and [`MongoQueryMethod`](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoQueryMethod.java#L197-L212).

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchHit.java:
```diff
@@ -0,0 +1,76 @@
+	private final float score;
+	private final T hit;
+
+	public static <T> Builder<T> builder() {
```
Let's avoid the builder pattern for simple objects such as this one. Having a public constructor is fine for a simple domain object. All of our [domain objects](https://github.com/spring-projects/spring-data-commons/blob/master/src/main/java/org/springframework/data/geo/GeoPage.java) in commons follow this idea which makes it easy to use and reuse such types.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchHit.java:
```diff
@@ -0,0 +1,76 @@
+	/**
+	 * @return the score for the hit.
+	 */
+	float getScore() {
```
Nit: `public`

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java:
```diff
@@ -95,14 +95,15 @@ public void delete(Query query, Class<?> clazz, IndexCoordinates index) {
 
 	@Override
-	public <T> Page<T> moreLikeThis(MoreLikeThisQuery query, Class<T> clazz, IndexCoordinates index) {
+	public <T> AggregatedPage<SearchHit<T>> moreLikeThis(MoreLikeThisQuery query, Class<T> clazz,
```
How does `AggregatedPage` relate to `SearchHit`/`SearchHits`? Can there be an `AggregatedPage` that contains `T` instead of `SearchHit<T>`?

Related (in the context of a new ticket): Would it make sense to rename `AggregatedPage` to `AggregationPage` and let a deprecated `AggregatedPage` extend from `AggregationPage`?

sothawo replied:
`AggregatedPage` extends `ScrolledPage>`and `ScoredPage` (I'm not sure if we need these as separate interfaces, as `AggregatedPageImpl` is the only implementaton of it all, and it adds ES aggregations. I don't like this class as it exposes ES objects in the API, but at the moment it's the only way to return aggregations. That's something I'd like to change later.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java:
```diff
@@ -135,9 +138,10 @@ public void delete(Query query, Class<?> clazz, IndexCoordinates index) {
+			MultiSearchRequest request) {
 		MultiSearchResponse.Item[] items = getMultiSearchResult(request);
-		List<Page<T>> res = new ArrayList<>(queries.size());
+		List<Page<SearchHit<T>>> res = new ArrayList<>(queries.size());
```
How about a new ticket that introduces a `MultiResult<T>` type? `List<Page<SearchHit<T>>>` is somewhat hard to read.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchHit.java:
```diff
@@ -0,0 +1,76 @@
+public final class SearchHit<T> {
+
+	private final float score;
+	private final T hit;
```
Nit: `T content`

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchHit.java:
```diff
@@ -0,0 +1,76 @@
+ * @author Peter-Josef Meisch
+ * @since 4.0
+ */
+public final class SearchHit<T> {
```
Let's also add a `toString` method, see [`GeoResult`](https://github.com/spring-projects/spring-data-commons/blob/master/src/main/java/org/springframework/data/geo/GeoResult.java#L44) for some inspiration.

Does it make sense to also associated the document Id with the `SearchHit` as the identifier has a fixed ES type?

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -81,7 +81,7 @@ default long count(Query query, IndexCoordinates index) {
 	 * @return a page with aggregations
 	 */
-	<T> AggregatedPage<T> queryForPage(Query query, Class<T> clazz, IndexCoordinates index);
+	<T> AggregatedPage<SearchHit<T>> queryForPage(Query query, Class<T> clazz, IndexCoordinates index);
```
Related to the question above: Changing the component type to `SearchHit<T>` could make it hard for users to consume the actual result. We might want to add convenience methods to `AggregatedPage` to consume a `List<T>` ( `List<T> unwrap()`?). Just a thought.

mp911de commented on src/test/java/org/springframework/data/elasticsearch/core/geo/ElasticsearchTemplateGeoTests.java:
```diff
@@ -172,7 +173,7 @@ public void shouldFindStringAnnotatedGeoMarkersInRangeForGivenCriteriaQuery() {
 				new Criteria("locationAsString").within(new GeoPoint(51.000000, 0.100000), "1km"));
 		// when
-		List<LocationMarkerEntity> geoAuthorsForGeoCriteria = operations.queryForList(geoLocationCriteriaQuery,
+		List<SearchHit<LocationMarkerEntity>> geoAuthorsForGeoCriteria = operations.queryForList(geoLocationCriteriaQuery,
```
Once we have `SearchHits`, the migration will be likely easier when users can switch from `List<T>` to `SearchHits<T>`.

mp911de commented on src/test/java/org/springframework/data/elasticsearch/core/query/CriteriaQueryTests.java:
```diff
@@ -777,7 +778,7 @@ public void shouldPerformBoostOperation() {
 
 		// when
-		Page<SampleEntity> page = operations.queryForPage(criteriaQuery, SampleEntity.class, index);
+		Page<SearchHit<SampleEntity>> page = operations.queryForPage(criteriaQuery, SampleEntity.class, index);
```
Would it also make sense to introduce a `SearchPage` type to avoid the deep generic parametrization?

sothawo commented:
I'll address the "minor" issues later. About the ideas with `SearchHits`, `SearchPage` etc: I agree, but I wouldn't wnat to do this all in this ticket. This should be the first, but definitely not the last step. So I'd prefer to first add the `SearchHit` - to the reactive part as well - and then in following tickets improve the API and user experience in smaller steps

sothawo commented:
as for different methods for returning `List<T>` or `List<SearchHit<T>>`: This goes somehow together with the repositories: There are in both reactive and non-reactive repositories the _find_ methods returning `List<T>` and `Flux<T>`. These we cannot change, this comes from spring-data-commons.

The reactive repositories up to now only use _find_ whereas the non-reactive use _find_ and _search_

The reactive operations use _find_, the non-reactive _query_.

Hard to find a way to have as little breaking change as possible.

xhaggi commented on src/main/java/org/springframework/data/elasticsearch/core/SearchHitSupport.java:
```diff
@@ -81,4 +88,55 @@ public static Object unwrapSearchHits(Object result) {
+	}
+
+	private static Pageable pageableOrUnpaged(Pageable pageable) {
+		return ofNullable(pageable).orElse(Pageable.unpaged());
```
Why not a simple null check?

sothawo replied:
I took this from `AggregatePageImpl` because it's a nice one-liner. The null check version would be:
```
	private static Pageable pageableOrUnpaged(Pageable pageable) {
		if(pageable != null) {
			return pageable;
		} else {
			return Pageable.unpaged();
		}
        }
```

xhaggi commented on src/main/java/org/springframework/data/elasticsearch/core/SearchHitSupport.java:
```diff
@@ -81,4 +88,55 @@ public static Object unwrapSearchHits(Object result) {
 		return result;
 	}
+
+	public static <T> SearchPage<T> searchPageFor(SearchHits<T> searchHits, @Nullable Pageable pageable) {
```
`toSearchPage()`, IMO sound better.

sothawo replied:
For me, methods named _toSomething()_ transform the object that they are called upon into something different. Here we have a static utility function so I deliberately did not use _to.._

xhaggi commented on src/main/java/org/springframework/data/elasticsearch/core/SearchPage.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * An aggregated page that holds a {@link SearchHits} element.
+ * 
+ * @author Peter-Josef Meisch
```
`@since 4.0`?

sothawo commented:
Got quite further than I thought. We now have:

- `SearchHit<T>` enriching  a T with _id_ and _score_
- `SearchHits<T>` for a search result, containing _totalHits_, _maxScore_,_scrollId_ and a list of `SearchHit<T>`.
- `SearchPage<T>` as a Page implementations with a `SearchHits<T>` and replacing the previous interfaces `AggregatePage`, `ScrolledPage` and `ScoredPage`.

The Operations and Templates now use these classes as search results; the repositories return the _normal_ `List<T>` and `Page<T>` as before.

This should give us a solid base to add more and more information from the search like highlights, sorts etc.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/repository/query/ElasticsearchStringQuery.java:
```diff
@@ -71,17 +72,23 @@ public Object execute(Object[] parameters) {
 		StringQuery stringQuery = createQuery(accessor);
```
In a later stage, we should move the query execution into a strategy object (see [`MongoQueryExecution`](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/MongoQueryExecution.java#L53) to execute the query according to what we want to see as results (see also [MongoDB](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/AbstractMongoQuery.java#L129-L156)).

Unconditionally calling `unwrapSearchHits` seems a workaround for a missing execution model.

sothawo replied:
Just a short reply before I find more time:

- I think the aggregations are up to now contained in the `AggregatedPage` because that was the only possibility to return additional information for a search result. From the ES concept, aggregations are an addition to a search result, so they should be moved to the `SearchHits` class (and I would not keep the `AggregatedPage`.
- Renaming from `query..()` to `search..()`: yes, but I would not mix this with this PR but do a follow up on this.
- I'll check the `scroll` methods and see how adding `SearchSlice` makes this cleaner, probably we should separate the scrolling from the paging.
- Totally replacing page with scroll would probably make sense as ES doesn't have the concept of pages and scrolling is faster than searching with offset and size, but that would have breaking changes for the repository interface as well, as up to now repositories can return pages and not the scrollid, and have to check if the scrollid is supported in the `Query`implementations.

Up to now the breaking changes are in the `SearchOperations` implementation with the repositories working as before; I'd like to keep that this way for now.

mp911de replied:
What do you think about reducing this PR to introducing the new `SearchHit`/`SearchHits`/… types only? I agree that the method changes are a better fit for a separate PR.

sothawo replied:
that would be commit f4203f9 plus adding some info to the `SearchHits` plus some more info to the class. I'll get that running and make further tickets for changing the methods, moving aggregates from the page to SearchHits etc.

xhaggi replied:
>Totally replacing page with scroll would probably make sense as ES doesn't have the concept of pages and scrolling is faster than searching with offset and size, but that would have breaking changes for the repository interface as well, as up to now repositories can return pages and not the scrollid, and have to check if the scrollid is supported in the Query implementations.

Scrolling is a stateful operation in ES, while searching is stateless. Using Scroll as the default for repo methods that return a page is not a good idea.

mp911de replied:
> Totally replacing page with scroll would probably make sense

I was referring to `startScroll` and `continueScroll` methods which return a `Page` object. Both should rather return a `Slice` type. Repository methods are not affected by this change.

xhaggi replied:
Ok sorry for the noise here.

sothawo commented:
I removed the changes that went beyond the `SearchHit`and `SearchHits`.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchHits.java:
```diff
@@ -0,0 +1,98 @@
+ * @author Peter-Josef Meisch
+ * @since 4.0
+ */
+public class SearchHits<T> implements Iterable<SearchHit<T>> {
```
We could implement `Streamable` instead of `Iterable` so we would inherit improved `Stream` support out of the box.

sothawo commented:
changed from `Iterable`to `Streamable and changed `SearchOperations` so the all the methods that return `SearchHit(s)` now have names starting with _search_ , the old methods are unchanged in their signature but deprecated and they now just use the _search_ methods.

Still need to introduce the new functions to the reactive part - a task for tomorrow.

sothawo commented:
now the `ReactiveSearchOperations` as well has the _find_ methods with the old return types deprecated and has _search_ methods returning the `SearchHit`variants.

That was a good idea to do it this way instead of renamimg them later, thx Mark

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param clazz The class of entity to retrieve.
+	 * @return scrolled page result
+	 */
+	<T> ScrolledPage<SearchHit<T>> searchScrollContinue(@Nullable String scrollId, long scrollTimeInMillis,
```
`startScroll`/`continueScroll` methods are rather internal methods and we expose a streaming functionality on top of these. Once we have `SearchSlice` in place, we should be good to just change the return types of `startScroll`/`continueScroll` so there's no need to have `searchScrollContinue`/`searchScrollClear`.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param index the index to run the query against
+	 * @return the first matching object
+	 */
+	default <T> SearchHit<T> searchForObject(Query query, Class<T> clazz, IndexCoordinates index) {
```
Let's stick with `searchOne` to align with the `findOne`/`selectOne` naming scheme that we have in the other modules.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param index the index to run the query against
+	 * @return list of found objects
+	 */
+	<T> SearchHits<T> searchForList(Query query, Class<T> clazz, IndexCoordinates index);
```
Let's stick with just `search` to align with Spring Data naming.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param index the index to run the query against
+	 * @return list of pages with the results
+	 */
+	<T> List<Page<SearchHit<T>>> searchForPage(List<? extends Query> queries, Class<T> clazz, IndexCoordinates index);
```
Let's rename this method to `multiSearch`. I also wonder what's the benefit of having `List<Page<…>>` over `List<SearchHits<T>>` in the return type.

sothawo replied:
The only need of having the page returned instead of the list is that this currently is the only possibility to get aggregations returned. Aggregations don't belong in the page but in the `SearchHits`. I have created a ticket for moving them there, so when we have that implemented, we can consolidate the different methods returning `List` and `Page` and remove the page variants because the page returned does not hold relevant info then. This affects not only _multiSearch_  but the single query _search_ as well.

sothawo replied:
I renamed it to `multiSearchForPage` and changed the overload that retruns a `List<SearchHits<T>>` to `multiSearch`. We should aim to keep the methods returning the `SearchHits` and not the ones with the `AggregatedPage`s once we have the necessary information in the `SearchHits`class.

mp911de replied:
Thanks for the detail. I also felt that aggregations should go into `SearchHits` (or a subtype of it).

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param index the index to run the query against
+	 * @return list of pages with the results
+	 */
+	List<AggregatedPage<? extends SearchHit<?>>> searchForPage(List<? extends Query> queries, List<Class<?>> classes,
```
Let's rename this method to `multiSearch`. I also wonder what's the benefit of having `List<Page<…>>` over `List<SearchHits<T>>` in the return type.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param <T> element return type
+	 * @return list of found objects
+	 */
+	default <T> List<List<SearchHit<T>>> searchForList(List<Query> queries, Class<T> clazz, IndexCoordinates index) {
```
If we rename `searchForPage(List, …)` to `multiSearch` and align the result to `List<SearchHits<T>>`, there's no need for this method, right?

sothawo replied:
I changed this to `List<SearchHits<T>> multiSearch(List<Query> queries, Class<T> clazz, IndexCoordinates index)`.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param index the index to run the query against
+	 * @return list of list of found objects
+	 */
+	default List<List<?>> searchForList(List<Query> queries, List<Class<?>> classes, IndexCoordinates index) {
```
If we rename `searchForPage(List, …)` to `multiSearch` and align the result to `List<SearchHits<T>>`, there's no need for this method, right?

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
 	 * @param index the index to run the query against
+	 * @return list of found object ids
+	 */
+	List<String> searchForIds(Query query, Class<?> clazz, IndexCoordinates index);
```
Let's not introduce this method as this method creates API surface for something that is easily consumable with `SearchHit.getId()`. Internally, there's no mapping happening and when using `SearchDocument`, we also skip mapping.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param index the index to run the query against
+	 * @return scrolled page result
+	 */
+	<T> ScrolledPage<SearchHit<T>> searchScrollStart(long scrollTimeInMillis, Query query, Class<T> clazz,
```
How about not introducing these methods yet as we should have `SearchSlice` in place first. See also above my comment that scroll methods have a rather internal flavor and we should be able to just change `startScroll`/`continueScroll` return types.

sothawo replied:
We need the `SearchHit` variant of these for the `public <T> CloseableIterator<SearchHit<T>> searchForStream(Query query, Class<T> clazz, IndexCoordinates index)`method to be able to return a stream of `SearchHit` objects. If we change the return of `startScroll/continueScroll` later, we cannot change the stream now.

mp911de replied:
Okay. Let's revisit the `search…Scroll` methods at a later stage, once we have all result and slice types in place.

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/convert/ElasticsearchConverter.java:
```diff
@@ -72,14 +75,24 @@ default ProjectionFactory getProjectionFactory() {
+	 * @param <T> the clazz of the type, must not be {@literal null}.
+	 * @return SearchHIt with all available information filled in
+	 */
+	<T> SearchHit<T> toSearchHit(SearchDocument searchDocument, Class<T> type);
```
Let's align with `EntityReader`'s `read` method so we would have `<R> SearchHit<R> read(Class<R> type, SearchDocument searchDocument);`

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +202,203 @@ default long count(Query query, IndexCoordinates index) {
+			IndexCoordinates index);
+
+	/**
+	 * Execute the criteria query against elasticsearch and return result as {@link List}
```
We should also cleanup the Javadoc before cleanup (`SearchHits` instead of list). Check also the other methods.

sothawo commented:
Did more refactorings and cleanups according to the review comments.

The scroll stuff I'd like to do together with a `SearchSlice`.

As for the `AggregatedPage` we won't need this anymore when aggregations are moved to the SearchHits; then the `ScoredPage` can be dropped as well as this info already in in the `SearchHits`.

Next steps would be:
- add sortValues from the result to `SearchHit`(implemented already, just waiting for this PR to be done)
- move aggregations to `SearchHits` and get rid of `AggregatedPage` and `ScoredPage`
- introduce `SearchSlice`and refactor scrolling / streaming

mp911de commented on src/main/java/org/springframework/data/elasticsearch/core/SearchOperations.java:
```diff
@@ -170,32 +203,204 @@ default long count(Query query, IndexCoordinates index) {
+	 * @param index the index to run the query against
+	 * @return the first found object
+	 */
+	default <T> SearchHit<T> searchOne(Query query, Class<T> clazz, IndexCoordinates index) {
```
Nit: `@Nullable` return.
